src/engine/SortedUnionImpl.h

```diff
@@ -0,0 +1,241 @@
+struct SortedUnionImpl
+    : ad_utility::InputRangeFromGet<Result::IdTableVocabPair> {
+  // Iterator and range storage.
+  std::shared_ptr<const Result> result1_;
+  std::shared_ptr<const Result> result2_;
+  Range1 range1_;
+  std::optional<typename Range1::iterator> it1_ = std::nullopt;
+  size_t index1_ = 0;
+  Range2 range2_;
+  std::optional<typename Range2::iterator> it2_ = std::nullopt;
+  size_t index2_ = 0;
+
+  // Result storage.
+  IdTable resultTable_;
+  LocalVocab localVocab_{};
+
+  // Metadata
+  ad_utility::AllocatorWithLimit<Id> allocator_;
+  bool requestLaziness_;
+  std::vector<std::array<size_t, 2>> columnOrigins_;
+  std::vector<ColumnIndex> leftPermutation_;
+  std::vector<ColumnIndex> rightPermutation_;
+  std::vector<std::array<size_t, 2>> targetOrder_;
+  bool aggregatedTableReturned_ = false;
+  // Function forwarded from `Union`
+  Func applyPermutation_;
+
+  SortedUnionImpl(std::shared_ptr<const Result> result1,
+                  std::shared_ptr<const Result> result2, Range1 range1,
+                  Range2 range2, bool requestLaziness,
+                  const std::vector<std::array<size_t, 2>>& columnOrigins,
+                  const ad_utility::AllocatorWithLimit<Id>& allocator,
+                  std::vector<ColumnIndex> leftPermutation,
+                  std::vector<ColumnIndex> rightPermutation,
+                  std::span<const ColumnIndex, SPAN_SIZE> comparatorView,
+                  Func applyPermutation)
+      : result1_{std::move(result1)},
+        result2_{std::move(result2)},
+        range1_{std::move(range1)},
+        range2_{std::move(range2)},
+        resultTable_{columnOrigins.size(), allocator},
+        allocator_{allocator},
+        requestLaziness_{requestLaziness},
+        columnOrigins_{columnOrigins},
+        leftPermutation_{std::move(leftPermutation)},
+        rightPermutation_{std::move(rightPermutation)},
+        applyPermutation_{std::move(applyPermutation)} {
+    if (requestLaziness) {
+      resultTable_.reserve(Union::chunkSize);
+    }
+    targetOrder_.reserve(comparatorView.size());
+    for (auto& col : comparatorView) {
+      targetOrder_.push_back(columnOrigins_.at(col));
+    }
+  }
+
+  // Always inline makes makes a huge difference on large datasets.
+  AD_ALWAYS_INLINE bool isSmaller(const auto& row1, const auto& row2) const {
+    using StaticRange = std::span<const std::array<size_t, 2>, SPAN_SIZE>;
+    for (auto [index1, index2] : StaticRange{targetOrder_}) {
+      if (index1 == Union::NO_COLUMN) {
+        return true;
+      }
+      if (index2 == Union::NO_COLUMN) {
+        return false;
+      }
+      if (row1[index1] != row2[index2]) {
+        return row1[index1] < row2[index2];
+      }
+    }
+    return false;
+  }
+
+  // Fetch the next element from the range, make a copy if it's a `Wrapper`.
+  // Otherwise move it. If the range is exhausted, return `std::nullopt`.
```

Comment:
This function by the way gets much less cryptic (in particular with the index)
if you group alll the iterator stuff for one side into a struct.